Impose maximum level restriction for BatchedTensors

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39859 Batching rule for torch.mul
* #39581 Add batching rule for torch.sum(tensor, dims)
* **#39580 Impose maximum level restriction for BatchedTensors**

We support 64 total levels. This is done so that we can represent lists
of levels as a bitset that fits into a single `int64_t` and is a
reasonable upper bound because we only support (physical) tensors of up
to 64 dimensions with vmap (see kVmapMaxTensorDims).

Test Plan:
`./build/bin/vmap_test`. One day we'll test this with the vmap Python
API.

Differential Revision: [D21929249](https://our.internmc.facebook.com/intern/diff/D21929249)

zou3519 committed:
Impose maximum level restriction for BatchedTensors

We support 64 total levels. This is done so that we can represent lists
of levels as a bitset that fits into a single `int64_t` and is a
reasonable upper bound because we only support (physical) tensors of up
to 64 dimensions with vmap (see kVmapMaxTensorDims).

Test Plan:
`./build/bin/vmap_test`. One day we'll test this with the vmap Python
API.

[ghstack-poisoned]

zou3519 committed:
Update on "Impose maximum level restriction for BatchedTensors"


We support 64 total levels. This is done so that we can represent lists
of levels as a bitset that fits into a single `int64_t` and is a
reasonable upper bound because we only support (physical) tensors of up
to 64 dimensions with vmap (see kVmapMaxTensorDims).

Test Plan:
`./build/bin/vmap_test`. One day we'll test this with the vmap Python
API.

[ghstack-poisoned]

ezyang commented:
A more robust test would ensure the boundary case (64 levels) works correctly, while 65 errors.

zou3519 commented:
> A more robust test would ensure the boundary case (64 levels) works correctly, while 65 errors.

True. I'll add that before landing.

zou3519 committed:
Update on "Impose maximum level restriction for BatchedTensors"


We support 64 total levels. This is done so that we can represent lists
of levels as a bitset that fits into a single `int64_t` and is a
reasonable upper bound because we only support (physical) tensors of up
to 64 dimensions with vmap (see kVmapMaxTensorDims).

Test Plan:
`./build/bin/vmap_test`. One day we'll test this with the vmap Python
API.

[ghstack-poisoned]

zou3519 committed:
Update on "Impose maximum level restriction for BatchedTensors"


We support 64 total levels. This is done so that we can represent lists
of levels as a bitset that fits into a single `int64_t` and is a
reasonable upper bound because we only support (physical) tensors of up
to 64 dimensions with vmap (see kVmapMaxTensorDims).

Test Plan:
`./build/bin/vmap_test`. One day we'll test this with the vmap Python
API.

Differential Revision: [D21929249](https://our.internmc.facebook.com/intern/diff/D21929249)

[ghstack-poisoned]

zou3519 committed:
Update on "Impose maximum level restriction for BatchedTensors"


We support 64 total levels. This is done so that we can represent lists
of levels as a bitset that fits into a single `int64_t` and is a
reasonable upper bound because we only support (physical) tensors of up
to 64 dimensions with vmap (see kVmapMaxTensorDims).

Test Plan:
`./build/bin/vmap_test`. One day we'll test this with the vmap Python
API.

Differential Revision: [D21929249](https://our.internmc.facebook.com/intern/diff/D21929249)

[ghstack-poisoned]

zou3519 committed:
Update on "Impose maximum level restriction for BatchedTensors"


We support 64 total levels. This is done so that we can represent lists
of levels as a bitset that fits into a single `int64_t` and is a
reasonable upper bound because we only support (physical) tensors of up
to 64 dimensions with vmap (see kVmapMaxTensorDims).

Test Plan:
`./build/bin/vmap_test`. One day we'll test this with the vmap Python
API.

Differential Revision: [D21929249](https://our.internmc.facebook.com/intern/diff/D21929249)

[ghstack-poisoned]
